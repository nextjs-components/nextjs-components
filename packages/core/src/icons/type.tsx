import React from "react";

import { useIconSize } from "../contexts/IconSizeContext";
import { Props } from "./props";

export default function Type({ color = "currentcolor", size }: Props) {
  const iconSize = useIconSize();
  return (
    <svg
      fill="none"
      height={size || iconSize.size}
      shapeRendering="geometricPrecision"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="1.5"
      viewBox="0 0 24 24"
      width={size || iconSize.size}
      style={{ color }}
    >
      <path d="M4 7V4h16v3"></path>
      <path d="M9 20h6"></path>
      <path d="M12 4v16"></path>
    </svg>
  );
}
