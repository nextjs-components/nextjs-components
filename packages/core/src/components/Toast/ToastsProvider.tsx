import React from "react";
import { createContext, useRef, useState } from "react";

type ToastType = "success" | "error" | "violet";
export interface IToast {
  height: number;
  key: string;
  preserve?: boolean;
  text: string;
  type?: ToastType;
  action?: string;
  cancelAction?: string;
}

export interface IToastsContext {
  current?: {
    messages: IToast[];
    message: (message: Partial<IToast> | string) => void;
    error: (message: Partial<IToast> | string) => void;
    success: (message: Partial<IToast> | string) => void;
    removeToastByKey: (key: string) => void;
  };
}

const throwerFn = () => {
  throw new Error("Missing a ToastsProvider");
};

export const ToastsContext = createContext<IToastsContext>({
  current: {
    messages: [],
    message: throwerFn,
    error: throwerFn,
    success: throwerFn,
    removeToastByKey: throwerFn,
  },
});

const ToastsProvider = ({ children }) => {
  const [messages, setMessages] = useState<IToast[]>([]);

  const message = (toast: Partial<IToast> | string) => {
    let text, preserve, action, cancelAction;
    if (typeof toast === "string") {
      text = toast;
    } else {
      ({ text, preserve, action, cancelAction } = toast);
    }
    setMessages((prev) => [
      ...prev,
      {
        text,
        preserve,
        height: action || cancelAction ? 72 : 64,
        key: (+new Date()).toString(),
        action,
        cancelAction,
      },
    ]);
  };
  const error = (toast: Partial<IToast> | string) => {
    let text, preserve, action, cancelAction;
    if (typeof toast === "string") {
      text = toast;
    } else {
      ({ text, preserve, action, cancelAction } = toast);
    }
    setMessages((prev) => [
      ...prev,
      {
        text,
        preserve,
        height: action || cancelAction ? 72 : 64,
        key: (+new Date()).toString(),
        action,
        cancelAction,
        type: "error",
      },
    ]);
  };
  const success = (toast: Partial<IToast> | string) => {
    let text, preserve, action, cancelAction;
    if (typeof toast === "string") {
      text = toast;
    } else {
      ({ text, preserve, action, cancelAction } = toast);
    }
    setMessages((prev) => [
      ...prev,
      {
        text,
        preserve,
        height: action || cancelAction ? 72 : 64,
        key: (+new Date()).toString(),
        action,
        cancelAction,
        type: "success",
      },
    ]);
  };
  const removeToastByKey = (key: string) => {
    setMessages((prev) => prev.filter((p) => p.key !== key));
  };

  return (
    <ToastsContext.Provider
      value={{
        current: {
          messages,
          message,
          error,
          success,
          removeToastByKey,
        },
      }}
    >
      {children}
    </ToastsContext.Provider>
  );
};

export default ToastsProvider;
