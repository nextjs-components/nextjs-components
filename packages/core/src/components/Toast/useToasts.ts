import { useContext } from "react";

import { ToastsContext } from "./ToastsProvider";

export default function useToasts() {
  return useContext(ToastsContext);
}
