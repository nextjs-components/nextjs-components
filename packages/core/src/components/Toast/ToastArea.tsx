import { Portal } from "@radix-ui/react-portal";
import clsx from "clsx";
import React from "react";
import { useReducer, useRef } from "react";
import { useHover } from "react-aria";

import ToastContainer from "./ToastContainer";
import styles from "./toasts.module.css";
import useToasts from "./useToasts";

function useForceUpdate(): () => void {
  return useReducer(() => ({}), {})[1];
}

const ToastArea = () => {
  const { hoverProps, isHovered } = useHover({});
  const { current } = useToasts();

  const messages = current?.messages ?? [];

  const heights = useRef([]);

  const forceUpdate = useForceUpdate();

  return (
    <Portal>
      <div
        {...hoverProps}
        className={clsx("toast-area", styles.toastArea, {
          [styles.multiple]: messages.length > 1,
        })}
      >
        {messages.map((e, i) => {
          return (
            <ToastContainer
              key={e.key}
              hovering={isHovered}
              position={messages.length - 1 - i}
              text={e.text}
              height={e.height}
              heights={heights.current}
              remove={() => {
                current?.removeToastByKey(e.key);
              }}
              type={e.type}
              preserve={e.preserve}
              action={e.action}
              cancelAction={e.cancelAction}
              id={e.key}
              onMount={forceUpdate}
            />
          );
        })}
      </div>
    </Portal>
  );
};

export default ToastArea;
