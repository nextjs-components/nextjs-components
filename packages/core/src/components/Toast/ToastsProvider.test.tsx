import { act, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import type { UserEvent } from "@testing-library/user-event/setup/setup";
import React from "react";

import ToastArea from "./ToastArea";
import ToastsProvider from "./ToastsProvider";
import useToasts from "./useToasts";

jest.useFakeTimers();

jest.mock("components/Button", () => ({
  __esModule: true,
  Button: (props) => <button {...props} />,
}));

describe("ToastConsumer", () => {
  let user: UserEvent;
  let ToastConsumer: React.ComponentType;
  let wrapper: React.ComponentType<React.PropsWithChildren>;

  beforeEach(() => {
    user = userEvent.setup({
      // https://testing-library.com/docs/user-event/options#advancetimers
      advanceTimers: jest.advanceTimersByTime,
    });
    wrapper = ({ children }) => (
      <ToastsProvider>
        {children}
        <ToastArea />
      </ToastsProvider>
    );
    ToastConsumer = () => {
      const toasts = useToasts();
      return (
        <button onClick={() => toasts.current?.message("This is a toast")}>
          Show toast
        </button>
      );
    };
  });

  // GIVEN
  it("should render without any errors", () => {
    // WHEN
    const { asFragment } = render(<ToastConsumer />, { wrapper });
    // THEN
    expect(asFragment()).toMatchSnapshot();
  });

  // GIVEN
  it("should open a toast", async () => {
    // WHEN
    const { baseElement, getByRole } = render(<ToastConsumer />, { wrapper });
    const button = getByRole("button");
    const toastArea = baseElement.getElementsByClassName("toast-area")[0];

    // THEN
    expect(toastArea.children).toHaveLength(0);

    // WHEN
    await act(async () => {
      await user.click(button);
      // advance timer so toasts can have unique IDs
      jest.advanceTimersByTime(10);
      await user.click(button);
    });

    // THEN
    expect(toastArea.children).toHaveLength(2);
  });

  // GIVEN
  it("should auto clear toasts", async () => {
    // WHEN
    const { baseElement } = render(<ToastConsumer />, { wrapper });
    const button = screen.getByRole("button");
    const toastArea = baseElement.getElementsByClassName("toast-area")[0];

    await act(async () => {
      await user.click(button);
    });

    // For some reason, this timer run needs to exist in a
    // separate act block for the `setTimeout` to actually
    // call its callback.
    act(() => {
      jest.advanceTimersByTime(4000);
    });

    // THEN
    expect(toastArea.children).toHaveLength(0);
  });

  // GIVEN
  it("should preserve toasts", async () => {
    // WHEN
    const ToastConsumer = () => {
      const toasts = useToasts();
      return (
        <button
          onClick={() =>
            toasts.current?.message({
              text: "This is a toast",
              preserve: true,
            })
          }
        >
          Show persistent toast
        </button>
      );
    };
    const { baseElement } = render(<ToastConsumer />, { wrapper });
    const button = screen.getByRole("button");
    const toastArea = baseElement.getElementsByClassName("toast-area")[0];

    // WHEN
    await act(async () => {
      await user.click(button);
      jest.advanceTimersByTime(10);
      await user.click(button);
    });

    // Run timers to clear non-preserve toasts
    act(() => {
      jest.advanceTimersByTime(10000);
    });

    // THEN
    expect(toastArea.children).toHaveLength(2);
  });

  // GIVEN
  it("should render action buttons", async () => {
    // WHEN
    const ToastConsumer = () => {
      const toasts = useToasts();
      return (
        <button
          onClick={() =>
            toasts.current.message({
              text: "This is a toast",
              action: "An action button!",
              cancelAction: "A cancel button!",
            })
          }
        >
          Show toast
        </button>
      );
    };
    render(<ToastConsumer />, { wrapper });
    const button = screen.getByRole("button");

    // WHEN
    await act(async () => {
      await user.click(button);
    });

    // THEN
    expect(screen.getByText("An action button!")).toBeInTheDocument();
    expect(screen.getByText("A cancel button!")).toBeInTheDocument();
  });
});
