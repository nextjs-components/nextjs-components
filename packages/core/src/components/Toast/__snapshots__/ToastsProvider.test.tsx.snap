// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastConsumer should render without any errors 1`] = `
<DocumentFragment>
  <button>
    Show toast
  </button>
</DocumentFragment>
`;
